js/scorm/wrapper.js

```diff
@@ -478,13 +466,46 @@ class ScormWrapper {
       return false;
     }
     this.scorm.get(property);
-    return !this.isUnsupportedErrorCode(this.scorm.debug.getCode());
+    const isSupported = !this.isUnsupportedLastError();
+    if (!isSupported) this.logUnsupported(property);
+    return isSupported;
+  }
+
+  isChildSupported(property) {
+    if (property.includes('_children')) return this.isSupported(property);
+    this.logger.debug(`ScormWrapper::isChildSupported: _property=${property}`);
+    if (this.finishCalled) {
+      this.logger.debug('ScormWrapper::isChildSupported: ignoring request as \'finish\' has been called');
+      return;
+    }
+    if (!this.lmsConnected) {
+      this.handleConnectionError();
+      return false;
+    }
+    const paths = property.split('.');
+    const element = paths.pop();
+    // remove last path if contains indexes
+    if (/^\d+$|^n$/.test(paths[paths.length - 1])) paths.pop();
+    const parentPath = paths.join('.');
+    const children = this.scorm.get(`${parentPath}._children`);
+    const isSupported = !this.isUnsupportedLastError() && children.includes(element);
+    if (!isSupported) this.logUnsupported(property);
+    return isSupported;
   }
 
```

Comment:
Could we add a helper function?

```suggestion
  setValueIfSupported(property, value) {
    if (!this.isChildSupported(property)) {
      this.logUnsupported(property);
      return;
    }
    
      this.setValue(property, value);
  }
```

Reply:
We could, but support for child elements are handled differently to others, so you would need awareness of the children models to ensure the correct supported checks are made before attempting to set the value. Please see https://github.com/adaptlearning/adapt-contrib-spoor/pull/306#discussion_r1450448779.

These changes could be made so you only need to call `setValue` and the wrapper takes care of it for you regardless of the data models being used, but it becomes messier as you have to create those hardcoded references to accepted elements etc., or we keep it simple and expect some knowledge of SCORM and what methods to use, as is the case currently.

Reply:
There are 7 instances of `if (this.isChildSupported(path)) this.setValue(path, value);` that could be just `this.setValueIfChildSupported(path, value);`
```js
  setValueIfChildSupported(property, value) {
    if (!this.isChildSupported(property)) return;
    this.setValue(property, value);
  }
  ```

Reply:
This is another example of a change that doesn't align with what was already in the file, but can be included. It wasn't a concern previously for the 3 examples where `isSupported` was being checked before setting a value. If going this route, I assume we should also include a `setValueIfSupported` for data models which don't have `_children`, i.e. should `cmi.comments` be used by an author?

My main hesitation for not including it previously was that I didn't want a method to look like you could throw any property at it and it will handle the call without issue. Someone without full understanding of SCORM could potentially look at the API and attempt to use that method outside the wrapper for properties without `_children`, or that don't exist, not realising that LMS errors will be raised. Not that it is a big deal to raise an LMS error, as previously I wanted to remove all checks to supported elements and just handle the LMS errors accordingly (https://github.com/adaptlearning/adapt-contrib-spoor/issues/302).

Maybe this isn't a concern and we don't need to account for something being used incorrectly outside of the plugin. It could be a simple case of changing the API to denote that not all methods are intended to be public and we use private/protected methods such as this to simplify calls inside the wrapper, where we know data models are being used correctly with the appropriate method?

Reply:
I think that doing what Cahir suggested, condensing the repeated behaviour into a single function call:
```js
  setValueIfSupported(property, value) {
    if (!this.isChildSupported(property)) return;
    this.setValue(property, value);
  }
```
But changing its name to `setValueIfChildSupported` is sufficiently clear to cover all of the raised concerns. If `isChildSupported` is clear enough, which I think it is, assuming some knowledge of SCORM and the relatively isolated nature of this API, then `setValueIfChildSupported` is clear enough.

It is also possible to add jsdoc comments to a function definitions with clarifying descriptions. I don't think that it's necessary in this case however.

I don't think we need to be overly concerned about developers making illegal calls to `setValue` or `isChildSupported` or `setValueIfChildSupported`.